ExitIntegrationTest.java

```diff
@@ -0,0 +1,176 @@
+package bcccp.carpark.exit;
+
+import bcccp.carpark.CarSensor;
+import bcccp.carpark.Carpark;
+import bcccp.carpark.Gate;
+import bcccp.carpark.IGate;
+import bcccp.tickets.adhoc.AdhocTicket;
+import bcccp.tickets.adhoc.AdhocTicketDAO;
+import bcccp.tickets.adhoc.IAdhocTicketDAO;
+import bcccp.tickets.season.ISeasonTicketDAO;
+import bcccp.tickets.season.SeasonTicket;
+import bcccp.tickets.season.SeasonTicketDAO;
+import bcccp.tickets.season.UsageRecord;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+import org.junit.Before;
+import org.junit.FixMethodOrder;
+import org.junit.Test;
+import org.junit.runners.MethodSorters;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
+/**
+ *
+ * @author James Nocevski
+ */
+@FixMethodOrder(MethodSorters.NAME_ASCENDING)
+public class ExitIntegrationTest {
+    
+    IGate testGate;
+    CarSensor testIS;
+    CarSensor testOS; 
+    ExitUI testUI;
+    Carpark testCarpark;
+    AdhocTicket  testAdhoc;
+    SeasonTicket testSeason;
+    IAdhocTicketDAO mockAdhocDAO;
+    ISeasonTicketDAO mockSeasonDAO;
+    ExitController testExit;
+    String seasonTicketId, cParkName;
+    UsageRecord mockRecord;
+    int cParkCap;
+    
+    @Before
+    public void setUp() {
+        cParkName = "test name";
+        cParkCap = 5;
+        mockAdhocDAO = mock(AdhocTicketDAO.class);
+        mockSeasonDAO = mock(SeasonTicketDAO.class);
+        mockRecord = mock(UsageRecord.class);
+        testCarpark = new Carpark(cParkName, cParkCap, mockAdhocDAO, mockSeasonDAO);
+        testGate = new Gate(0, 0);
+        testIS = new CarSensor("test inside", 100, 100);
+        testOS = new CarSensor("test outside", 200, 200);
+        testUI = mock(ExitUI.class);
+        testAdhoc = new AdhocTicket(cParkName, 1, "Atest");
+        testSeason = new SeasonTicket("test", cParkName, 1234, 4321);
+        testExit = new ExitController(testCarpark, testGate, testIS, testOS, testUI);
+    }
+    
+    @Test
+    public void initialisationTest() {
+        assertTrue(testCarpark instanceof Carpark);
+        assertTrue(testIS instanceof CarSensor);
+        assertTrue(testOS instanceof CarSensor);
+        assertTrue(testExit instanceof ExitController);
+    }
+    
+    @Test
+    public void testNormalFlowSeason() {
+        assertEquals(testExit.getState(), "IDLE");  //checks initial state
+        testIS.setCarDetected(true);
+        testExit.carEventDetected(testIS.getId(), true);
+        
+        assertEquals(testExit.getState(), "WAITING");
+        verify(testUI).display("Insert Ticket");
+        
+        when(mockSeasonDAO.findTicketById(testSeason.getId())).thenReturn(testSeason);
+        testSeason.recordUsage(mockRecord);
+        testExit.ticketInserted(testSeason.getId());
+        assertEquals(testExit.getState(), "PROCESSED");

```

Comment:
here, after the ticket is inserted, you need to check the system followed the use case
4.1 System ensures season ticket valid and in use
you need some check to see those methods were called and the right values returned

Reply:
Verification is done by ExitController, if everything was called and returned correctly, the state is set to PROCESSED, so that's what I've tested for, sorry for the confusion.

Reply:
it is not reasonable to use an outcome as evidence that everything leading up to that outcome happened as expected. 
You have to prove each step along the way and then finally the outcome. If your test failed, would you know what had gone wrong? Could you say it failed because the season ticket was not valid, or because it was not in use or because the system failed to check one of those things correctly or because setState is reading the wrong variables?
I don't think only testing the inital conditions and the end conditions and saying that everything in the middle must be ok is thorough enough.

Reply:
Fair point, I have added asserts that test that the methods that are called return the right values.